samples-check/src/main/java/org/gradle/exemplar/test/engine/SamplesRunnerJUnitEngine.java

```diff
@@ -0,0 +1,313 @@
+            final List<? extends OutputNormalizer> normalizers
+    ) {
+        if (command.getExpectedOutput() == null) {
+            return;
+        }
+
+        String expectedOutput = command.getExpectedOutput();
+        String actualOutput = executionResult.getOutput();
+
+        for (OutputNormalizer normalizer : normalizers) {
+            actualOutput = normalizer.normalize(actualOutput, executionResult.getExecutionMetadata());
+        }
+
+        if (command.isAllowDisorderedOutput()) {
+            new AnyOrderLineSegmentedOutputVerifier().verify(expectedOutput, actualOutput, command.isAllowAdditionalOutput());
+        } else {
+            new StrictOrderLineSegmentedOutputVerifier().verify(expectedOutput, actualOutput, command.isAllowAdditionalOutput());
+        }
+    }
+
+    private CommandExecutionResult executeSample(SampleTestDescriptor descriptor, File workingDir, Command command) {
+        ExecutionMetadata executionMetadata = getExecutionMetadata(descriptor.sample.getProjectDir());
+        CommandExecutorParams commandExecutorParams = new CommandExecutorParams(executionMetadata, workingDir, command);
+        return descriptor.commandExecutorFunction
+                .apply(commandExecutorParams)
+                .execute(command, executionMetadata);
+    }
+
+    private ExecutionMetadata getExecutionMetadata(final File tempSampleOutputDir) {
+        Map<String, String> systemProperties = new HashMap<>();
+        for (String systemPropertyKey : SAFE_SYSTEM_PROPERTIES) {
+            systemProperties.put(systemPropertyKey, System.getProperty(systemPropertyKey));
+        }
+
+        return new ExecutionMetadata(tempSampleOutputDir, systemProperties);
+    }
+
+    private Sample initSample(final SampleTestDescriptor descriptor) throws IOException {
+        Sample sampleIn = descriptor.sample;
+        File tmpProjectDir = new File(tempDir, sampleIn.getId());
+        FileUtils.copyDirectory(sampleIn.getProjectDir(), tmpProjectDir);
+        Sample sample = new Sample(sampleIn.getId(), tmpProjectDir, sampleIn.getCommands());
+        for (SampleModifier sampleModifier : descriptor.sampleModifiers) {
+            sample = sampleModifier.modify(sample);
+        }
+        return sample;
+    }
+
+    private static class SampleTestDescriptor extends AbstractTestDescriptor {
+
+        private final Sample sample;
+        private final Function<CommandExecutorParams, CommandExecutor> commandExecutorFunction;
+        private final List<? extends OutputNormalizer> normalizers;
+        private final List<SampleModifier> sampleModifiers;
+
+        private SampleTestDescriptor(
+                UniqueId uniqueId,
+                Sample sample,
+                Function<CommandExecutorParams, CommandExecutor> commandExecutorFunction, List<? extends OutputNormalizer> normalizers,
+                List<SampleModifier> sampleModifiers
+        ) {
+            super(uniqueId, sample.getId());
```

Comment:
We should pass a `FileSource` here so that one can navigate to the source in IDEs.